Add sanity test for multi DC and several improvements for test.py

xtrey opened the pull request:
Add sanity tests for multi dc

Fix writing cassandra-rackdc.properties with correct format data instead of yaml
Add a parameter to overwrite RF for specific DC
Add the possibility to connect cql to the specific node


In this PR 4 tests were added to test multi-DC functionality. One is added from initial commit were multi-DC possibility were introduced, however, this test was not commited. Three of them are migrations from dtest, that later will be deleted. To be able to execute migrated tests additional functionality is added: the ability to connect cql to the specific node in the cluster instead of pooled connection and the possibility to overwrite the replication factor for the specific DC. To be able to use the multi DC in test.py issue with the incorrect format of the properties file fixed in this PR.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6972/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6972/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6972/)
#### Failed Tests (304/23408):
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_multidc/test_multidc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_putget_2dc_rf2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_multidc/test_putget_2dc_rf2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_putget_2dc_rf2)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_multidc/test_multidc_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_putget_2dc_rf1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_multidc/test_putget_2dc_rf1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_putget_2dc_rf1)
- [test_putget_2dc_rf2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_multidc/test_putget_2dc_rf2_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_putget_2dc_rf2)
- [test_replace_with_same_ip_twice](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_replace_with_same_ip_twice/test_replace_with_same_ip_twice) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice)
- [test_sticky_coordinator_enforced](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_select_from_mutation_fragments/test_sticky_coordinator_enforced) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sticky_coordinator_enforced)
- [test_tablet_replication_factor_enough_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_tablets/test_tablet_replication_factor_enough_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_replication_factor_enough_nodes)
- [test_tablet_cannot_decommision_below_replication_factor](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_tablets/test_tablet_cannot_decommision_below_replication_factor) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cannot_decommision_below_replication_factor)
- [test_tablet_replication_factor_enough_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6972/testReport/junit/%28root%29/test_tablets/test_tablet_replication_factor_enough_nodes_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_replication_factor_enough_nodes)
#### Build Details:
- Duration: 2 hr 32 min
- Builder: spider1.cloudius-systems.com

enaydanov commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,207 @@
+import sys
+sys.path.insert(0, sys.path[0] + "/test/cql-pytest")
+import nodetool
+import logging
+import time
```
These are libraries from the stdlib, they should be in a first block separated from the 3rd-party libraries by one empty line.

enaydanov commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,207 @@
+    cf = "test_cf"
+    table_name = "test_table_name"
+    columns = [Column("name", TextType), Column("value", TextType)]
+    for i in [1, 2]:
+        s_info = await manager.server_add(
+            config={"snitch": "GossipingPropertyFileSnitch"},
+            property_file={"dc": "dc{}".format(i), "rack": "myrack"},
+        )
+        logger.info(s_info)
```
wdyt about following loop:

```python
for i in [1, 1, 2, 2]:
    ...
```

enaydanov commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,207 @@
+    return new_list
+
+
+def _put_with_overwrite(session, nb_keys, cf, ks, cl=ConsistencyLevel.QUORUM):
```
Need to add some explanation about records you put to DB here.

enaydanov commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,207 @@
+    _validate_row(rows)
+
+
+def _validate_row(res):
```
Without any comments it looks like some magic : )

enaydanov reviewed (changes requested):
In addition:
1. Why you need 2 commits in this PR?
2. Please, add description to the commit/commits and to the PR
3. Add more logging in tests.

xtrey commented:
> In addition:
> 
> 1. Why you need 2 commits in this PR?
> 2. Please, add description to the commit/commits and to the PR
> 3. Add more logging in tests.

1. No, there is no need in several commit. Fixed
2. Added more information
3. Additional logging added.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7099/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7099/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7099/)
#### Failed Tests (402/23516):
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_multidc/test_multidc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_multidc/test_multidc_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_putget_2dc_rf2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_multidc/test_putget_2dc_rf2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_putget_2dc_rf2)
- [test_query_dc_with_rf_0_does_not_crash_db](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db)
- [test_replace_with_same_ip_twice](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_replace_with_same_ip_twice/test_replace_with_same_ip_twice) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_with_same_ip_twice)
- [test_sticky_coordinator_enforced](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_select_from_mutation_fragments/test_sticky_coordinator_enforced) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sticky_coordinator_enforced)
- [test_sticky_coordinator_enforced](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_select_from_mutation_fragments/test_sticky_coordinator_enforced_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_sticky_coordinator_enforced)
- [test_drop_table_during_streaming_receiver_side](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_table_drop/test_drop_table_during_streaming_receiver_side) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_drop_table_during_streaming_receiver_side)
- [test_tablet_replication_factor_enough_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_tablets/test_tablet_replication_factor_enough_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_replication_factor_enough_nodes)
- [test_tablet_cannot_decommision_below_replication_factor](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7099/testReport/junit/%28root%29/test_tablets/test_tablet_cannot_decommision_below_replication_factor) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cannot_decommision_below_replication_factor)
#### Build Details:
- Duration: 2 hr 26 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7108/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7108/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7108/)
#### Failed Tests (39/23516):
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_multidc/test_multidc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_remove_garbage_group0_members](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_topology_remove_garbage_group0/test_remove_garbage_group0_members) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_group0_members)
- [test_remove_garbage_group0_members](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_topology_remove_garbage_group0/test_remove_garbage_group0_members_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_group0_members)
- [test_nodes_with_different_smp](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_topology_smp/test_nodes_with_different_smp) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nodes_with_different_smp)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_multidc/test_multidc_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_multidc/test_multidc_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_multidc/test_multidc_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_remove_garbage_group0_members](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_topology_remove_garbage_group0/test_remove_garbage_group0_members_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_group0_members)
- [test_remove_garbage_group0_members](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_topology_remove_garbage_group0/test_remove_garbage_group0_members_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_remove_garbage_group0_members)
- [test_nodes_with_different_smp](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7108/testReport/junit/%28root%29/test_topology_smp/test_nodes_with_different_smp_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_nodes_with_different_smp)
#### Build Details:
- Duration: 2 hr 20 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7107/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7107/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/246/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8284/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7107/)
#### Failed Tests (45/23516):
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_multidc/test_multidc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_rolling_upgrade_happy_path](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_raft_cluster_features/test_rolling_upgrade_happy_path) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade_happy_path)
- [test_rolling_upgrade_happy_path](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_raft_cluster_features/test_rolling_upgrade_happy_path_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_rolling_upgrade_happy_path)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_multidc/test_multidc_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_multidc/test_multidc_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_putget_2dc_rf2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_multidc/test_putget_2dc_rf2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_putget_2dc_rf2)
- [test_query_dc_with_rf_0_does_not_crash_db](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_multidc/test_query_dc_with_rf_0_does_not_crash_db) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_dc_with_rf_0_does_not_crash_db)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_multidc/test_multidc_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_no_removed_node_event_on_ip_change](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_no_removed_node_event_on_ip_change/test_no_removed_node_event_on_ip_change) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_no_removed_node_event_on_ip_change)
- [test_node_shutdown_waits_for_pending_requests](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7107/testReport/junit/%28root%29/test_node_shutdown_waits_for_pending_requests/test_node_shutdown_waits_for_pending_requests) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_node_shutdown_waits_for_pending_requests)
#### Build Details:
- Duration: 4 hr 43 min
- Builder: spider8.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7129/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7129/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7129/)
#### Failed Tests (792/23516):
- [topology_custom.test_multidc.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_multidc_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_multidc.debug.1)
- [topology_custom.test_multidc.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_multidc_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_multidc.debug.2)
- [topology_custom.test_multidc.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_multidc_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_multidc.debug.3)
- [topology_custom.test_replace_with_same_ip_twice.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_replace_with_same_ip_twice_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_replace_with_same_ip_twice.debug.1)
- [topology_custom.test_replace_with_same_ip_twice.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_replace_with_same_ip_twice_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_replace_with_same_ip_twice.debug.2)
- [topology_custom.test_replace_with_same_ip_twice.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_replace_with_same_ip_twice_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_replace_with_same_ip_twice.debug.3)
- [topology_custom.test_select_from_mutation_fragments.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_select_from_mutation_fragments_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_select_from_mutation_fragments.debug.1)
- [topology_custom.test_select_from_mutation_fragments.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_select_from_mutation_fragments_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_select_from_mutation_fragments.debug.2)
- [topology_custom.test_select_from_mutation_fragments.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_select_from_mutation_fragments_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_select_from_mutation_fragments.debug.3)
- [topology_custom.test_table_drop.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7129/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_table_drop_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_table_drop.debug.1)
#### Build Details:
- Duration: 1 hr 51 min
- Builder: spider5.cloudius-systems.com

yaronkaikov commented:
@xtrey @enaydanov i am running the CI again https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7143/ this time it will run on fleet (we have different value for `aio-max-nr` ) , let see if it pass, i will align all local nodes as well

xtrey commented:
@yaronkaikov thanks. Looks like change is helped. Only one test failed, I'll investigate it.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7147/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7147/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7147/)
#### Build Details:
- Duration: 1 hr 55 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: ABORTED

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7166/)
#### Build Details:
- Duration: 16 min
- Builder: i-08d1e5429a740bdb5 (m5ad.8xlarge)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7167/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7167/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7167/)
#### Build Details:
- Duration: 1 hr 51 min
- Builder: spider4.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7200/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7200/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7200/)
#### Failed Tests (20/23599):
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_4) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_5) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_7) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_8) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [test_multidc](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/%28empty%29/test_multidc_9) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc)
- [topology_custom.test_multidc.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7200/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_multidc_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_multidc.debug.1)
#### Build Details:
- Duration: 1 hr 45 min
- Builder: spider7.cloudius-systems.com

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7207/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7207/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7207/)
#### Failed Tests (2/23734):
- [test_replace_ignore_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7207/testReport/junit/%28root%29/test_replace_ignore_nodes/test_replace_ignore_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_ignore_nodes)
- [topology_custom.test_replace_ignore_nodes.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7207/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_replace_ignore_nodes_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_replace_ignore_nodes.dev.3)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider1.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7248/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7248/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7248/)
#### Build Details:
- Duration: 1 hr 44 min
- Builder: spider2.cloudius-systems.com

nyh commented on test/pylib/random_tables.py:
```diff
@@ -256,16 +256,18 @@ def __str__(self):
-        self.manager.cql.execute(f"CREATE KEYSPACE {keyspace} WITH REPLICATION = "
-                                 "{ 'class' : 'NetworkTopologyStrategy', "
-                                 f"'replication_factor' : {replication_factor} }}")
+        self.keyspace_query = f"CREATE KEYSPACE {keyspace} WITH REPLICATION = {{ 'class' : 'NetworkTopologyStrategy', 'replication_factor' : {replication_factor}}}"
```
Why do you want to put this variable in "self"?  Do you plan to use it anywhere except in this function?

xtrey replied:
The idea here just not to repeat the same query twice. It used later in modifying the the replication factor for each DC in lines 267-270

nyh replied:
Yes, but all of these are **inside** the function, so this could have been a local variable, not an attribute of "self". Unless I'm missing something...

nyh replied:
By the way, @kostja maybe you want to review these changes to test/pylib so it doesn't "rot" in a way you don't like?

kostja replied:
I partly wrote them so I should be disqualified here. But I am OK if @xtrey submits the code that might "rot" a bit - he will bear the responsibility for the outcome, since I hope going forward he will be handling all test.py issues.
Thanks for pinging though.

xtrey replied:
@nyh sorry, misunderstood you. You are completely right about moving it to the local variable.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+"""
+Test functionality on the cluster with different values of the --smp parameter on the nodes.
```
I don't understand the relevance of this comment. Maybe it's a copypasto?

xtrey replied:
Correct, this is a copypaste from the https://github.com/kostja/scylla/tree/multidc2. There was one test where @kostja  made the initial commit to support multi-DC. I'll delete this comment to remove confusion.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+@pytest.mark.asyncio
+@pytest.mark.skip(reason="Too heavy for CI, but OK for local run")
+async def test_multidc(request: pytest.FixtureRequest, manager: ManagerClient) -> None:
+    logger.info("Creating a new node")
```
It looks like you are adding now "a new node", but 30 of them. 
Is there any benefit to adding 30 and not 10 or 3?

xtrey replied:
This test is copypaste. I do believe the idea of this test is to check that ScyllaDB will be able to handle a large amount of DC. Each node in this test is added to its own DC, so the test generally checks the ability to work with 30 DC.

nyh replied:
copypaste from what?
If the intention is to check a large number of DCs, this should be specified in the test's comment, and perhaps even the test's name (e.g., call it many_dc instead of multi_dc). But then you need to ask yourself, is 30 really many? What kind of datastructure might we have in Scylla which can deal with 10 DCs fine, but 30 DCs is already "many"?

bhalevy replied:
5 DCs is plenty, and atypical

mykaul replied:
One of our largest customers uses 5 DCs...

xtrey replied:
Changed to 5.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+CONFIG = {"endpoint_snitch": "GossipingPropertyFileSnitch"}
+
+
+# Checks a cluster boot/operations in multi-dc environment
```
This comment doesn't really explain what this test is about. First of all, it's not just a "mult-dc environment", it's a particular setup where, if I understand correctly, you have 30 nodes in 30 different DCs.
Second, what does this test actually check? That the 30 nodes are able to boot (how is boot different in the case of 30 different CDs and one DC)? That you are able to create a table in a 30-DC setup? I don't see any data operations below, only the table creation, so maybe you can explain why it's interesting.

kostja replied:
yes, the idea is to explore a rare dimension - 30-dc setup . If there is any quadratic or cubic logic in scylla from the number of DCs the test should uncover that. calling random_tables() would cover for DML and DDL in such a cluster. I don't know the actual run time of this test now, if it's too big, we could perhaps reduce the number of DCs.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+async def test_putget_2dc_rf1(
+        request: pytest.FixtureRequest, manager: ManagerClient
+) -> None:
+    """Simple put-get test for 2 DC with one node each (RF=1) [catches #3539]"""
```
https://github.com/scylladb/scylladb/issues/3539 doesn't seem related. So which issue did you mean?

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+
+
+@pytest.mark.asyncio
+async def test_putget_2dc_rf2(
```
Nitpick: It seems this test is identical to the previous one except for the array, so you could have done one copy of the test with "pytest.parameterize".

Another benefit is that you won't need the extra putget() function (which I, as a reader, have absolutely no idea what it does), you can insert it inside this function and let the reader read the test sequentially without jumping around the file.

xtrey replied:
Good point. I'll rewrite it to be parametrized.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+
+
+@pytest.mark.asyncio
+async def test_query_dc_with_rf_0_does_not_crash_db(request: pytest.FixtureRequest, manager: ManagerClient):
```
Another thing you forgot to mention in the commit message is what tests you are adding, and why. This particular test seems a direct "borrowing" of the existing dtest multidc_additional_tests.py. I think it was worth mentioning. And it may be worth considering (please do!) deleting this dtest later - we don't need exactly the same test twice.

Are other tests added here also translations of existing dtests?

xtrey replied:
All tests presented here are borrowed: 3 from dtests and 1 from https://github.com/kostja/scylla/blob/multidc2/test/topology_custom/test_multidc.py
Sure, I'll delete this tests from dtest once this PR will be merged. You are completely right there is no need to have the same tests in different repos.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+    query_cl_local_quorum = SimpleStatement(f"SELECT * from {ks}.{table_name};",
+                                            consistency_level=ConsistencyLevel.LOCAL_QUORUM)
+    nodetool.flush(dc1_connection, "{ks}.{table_name}")
+    first_node_results = list(dc1_connection.execute(query_cl_one).one())
```
I know this test is copied from the existin dtest, so I can't complain, but I'm still curious why it was important to query DC1 with CL=ONE... If the bug we were hunting was about LOCAL_QUORUM on DC2, doesn't it make sense to also query DC1 with LOCAL_QUORUM?

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+
+def _put_with_overwrite(session, nb_keys, cf, ks, cl=ConsistencyLevel.QUORUM):
+    """
+    Put 100 rows with for specified keys. Then overwrite partially existing rows with new values
```
It seems the code below actually adds nb_keys * 100 rows?
What is nb_keys about at all?

Personally, I would love for this function to disappear. If I understand correctly, you only use it once in the code. So just insert it where its needed, and add a comment *exactly* what it is writing.

Also, is all this elaborate code to write 100 rows and then rewrite some of them, etc., needed for this test? Does it test anything that just writing 10 rows once wouldn't have tested?

xtrey replied:
Modified and simplified code.
This is a migrated test, and I think the idea behind it is to check that there are no issues with updating the values in a multi-DC environment.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+
+    query = SimpleStatement(f"SELECT * FROM {cf} WHERE key='k0'", consistency_level=cl)
+    rows = list(session.execute(query))
+    _validate_row(rows)
```
Again, I think (but it's a matter of style, I guess) that reading this code would have been much simpler without all these functions and just one linear code which writes a bunch of data, reads it back and checks if it has the right value. It will be a long function - but clearer, could be read in a linear fashion (one line after the other) and no need to name functions, parameters (like that nb_keys which is always 1 anyway), etc.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+
+
+def putget(session, cf, ks, cl=ConsistencyLevel.QUORUM):
+    _put_with_overwrite(session, 1, cf, ks, cl)
```
So after the above generic function with an "nb_keys" parameter, you end up calling it just once, with nb_keys=1? :-( Can you get rid of this parameter and make everything easier to understand?

xtrey replied:
Removed unnecessary parameter

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,209 @@
+
+def _validate_row(res):
+    """This method checks final state of the table after overwrites done in :py:_put_with_overwrite():"""
+    assert_length_equal(res, 100)
```
Adding this assert_length_equal convenience function  is completely unnecessary and makes it harder to read the code. The idiomatic pytest way to do this is simply:
```
    assert len(res) == 100
```
And that's it! No need for an error string (assertion failures are very clear in pytest) and definitely not for a wrapper function for this assert.

xtrey replied:
Done

xtrey commented on test/pylib/manager_client.py:
```diff
@@ -81,6 +83,13 @@ def get_cql(self) -> CassandraSession:
 
+    @staticmethod
+    def connect_to_node(server: ServerInfo) -> CassandraSession:
+        """Connect cqlsh to specific node in the cluster"""
```
Changed to CQL instead of CQLSH. 

Sure, will modify the commit message when I'll resolve all comments.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7284/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7284/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7284/)
#### Build Details:
- Duration: 2 hr 2 min
- Builder: spider1.cloudius-systems.com

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,140 @@
+cluster_config = [
+    ([1, 2], 1),
+    ([1, 1, 2, 2], 2)
+]
```
Please move cluster_config to right above the place it is used, it will be clearer.

nyh commented on test/topology_custom/test_multidc.py:
```diff
@@ -0,0 +1,140 @@
+async def test_putget_2dc_with_rf(
+        request: pytest.FixtureRequest, manager: ManagerClient, nodes_list: list[int], rf: int
+) -> None:
+    """Simple put-get test for 2 DC with one node each"""
```
This comment is no longer accurate - it's just one of the two cluster_config options.

nyh reviewed (changes requested):
I like this version much better, thanks. Left just a few minor nitpicks.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7330/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7330/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7330/)
#### Failed Tests (2/23959):
- [test_replace_ignore_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7330/testReport/junit/%28root%29/test_replace_ignore_nodes/test_replace_ignore_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_ignore_nodes)
- [topology_custom.test_replace_ignore_nodes.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/7330/testReport/junit/%28root%29/non-boost%20tests/topology_custom_test_replace_ignore_nodes_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_replace_ignore_nodes.dev.1)
#### Build Details:
- Duration: 1 hr 50 min
- Builder: spider2.cloudius-systems.com

nyh reviewed (approved):
There is still one silly question I had about why a certain variable had to be a class variable (instead of a local variable) that I didn't get an answer to. The rest of this PR looks good.
I see the CI is still running so I can't merge it yet anyway.

xtrey commented:
> There is still one silly question I had about why a certain variable had to be a class variable (instead of a local variable) that I didn't get an answer to. The rest of this PR looks good. I see the CI is still running so I can't merge it yet anyway.

My apologies, missed your reply regarding this. Fixed.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7434/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7434/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7434/)
#### Build Details:
- Duration: 2 hr 4 min
- Builder: spider5.cloudius-systems.com

nyh commented:
@xtrey I already merged this PR two hours before you pushed newer versions. Just now it was even promoted to master. Please send an incremental patch of whatever you changed in your last push (I don't know what it is because you didn't write an explanation what you changed in the last push).

kbr-scylla commented:
The new test is flaky.

```
>       dc1_connection = manager.connect_to_node(servers[0])

test/topology_custom/test_multidc.py:126: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/pylib/manager_client.py:111: in connect_to_node
    return CassandraCluster([server.ip_addr], execution_profiles={EXEC_PROFILE_DEFAULT: profile}).connect()
cassandra/cluster.py:1761: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:1787: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:1774: in cassandra.cluster.Cluster.connect
    ???
cassandra/cluster.py:3603: in cassandra.cluster.ControlConnection.connect
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.cluster.NoHostAvailable: ('Unable to connect to any servers', {'127.112.126.21:9042': OperationTimedOut('errors=None, last_host=None')})
```

It's probably because you did not set increased timeouts like we do e.g. in topology/conftest.py for `cluster_con`:
```python3
def cluster_con(hosts: List[IPAddress | EndPoint], port: int, use_ssl: bool, auth_provider=None, load_balancing_policy=RoundRobinPolicy()):
    """Create a CQL Cluster connection object according to configuration.
       It does not .connect() yet."""
    assert len(hosts) > 0, "python driver connection needs at least one host to connect to"
    profile = ExecutionProfile(
        load_balancing_policy=load_balancing_policy,
        consistency_level=ConsistencyLevel.LOCAL_QUORUM,
        serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL,
        # The default timeouts should have been more than enough, but in some
        # extreme cases with a very slow debug build running on a slow or very busy
        # machine, they may not be. Observed tests reach 160 seconds. So it's
        # incremented to 200 seconds.
        # See issue #11289.
        # NOTE: request_timeout is the main cause of timeouts, even if logs say heartbeat
        request_timeout=200)
    whitelist_profile = ExecutionProfile(
        load_balancing_policy=TokenAwarePolicy(WhiteListRoundRobinPolicy(hosts)),
        consistency_level=ConsistencyLevel.LOCAL_QUORUM,
        serial_consistency_level=ConsistencyLevel.LOCAL_SERIAL,
        request_timeout=200)
    if use_ssl:
        # Scylla does not support any earlier TLS protocol. If you try,
        # you will get mysterious EOF errors (see issue #6971) :-(
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
    else:
        ssl_context = None

    return Cluster(execution_profiles={EXEC_PROFILE_DEFAULT: profile, 'whitelist': whitelist_profile},
                   contact_points=hosts,
                   port=port,
                   # TODO: make the protocol version an option, to allow testing with
                   # different versions. If we drop this setting completely, it will
                   # mean pick the latest version supported by the client and the server.
                   protocol_version=4,
                   # NOTE: No auth provider as auth keysppace has RF=1 and topology will take
                   # down nodes, causing errors. If auth is needed in the future for topology
                   # tests, they should bump up auth RF and run repair.
                   ssl_context=ssl_context,
                   # The default timeouts should have been more than enough, but in some
                   # extreme cases with a very slow debug build running on a slow or very busy
                   # machine, they may not be. Observed tests reach 160 seconds. So it's
                   # incremented to 200 seconds.
                   # See issue #11289.
                   connect_timeout = 200,
                   control_connection_timeout = 200,
                   # NOTE: max_schema_agreement_wait must be 2x or 3x smaller than request_timeout
                   # else the driver can't handle a server being down
                   max_schema_agreement_wait=20,
                   idle_heartbeat_timeout=200,
                   # The default reconnection policy has a large maximum interval
                   # between retries (600 seconds). In tests that restart/replace nodes,
                   # where a node can be unavailable for an extended period of time,
                   # this can cause the reconnection retry interval to get very large,
                   # longer than a test timeout.
                   reconnection_policy = ExponentialReconnectionPolicy(1.0, 4.0),

                   auth_provider=auth_provider,
                   # Capture messages for debugging purposes.
                   connection_class=CustomConnection
                   )
```

BTW why do we need a **yet another** method to connect to clusters?

kbr-scylla commented:
Opened https://github.com/scylladb/scylladb/issues/17914

xtrey commented:
@kbr-scylla you are right. There is no need for an additional method. The fix is there https://github.com/scylladb/scylladb/pull/17932

nyh commented:
I just fell on the cassandra-rackdc.properties formatting bug when backporting some test to an older branch (https://github.com/scylladb/scylla-enterprise/pull/4934).
